src/lib/network_pool/transaction_pool.ml

```diff
@@ -1052,17 +1135,60 @@ struct
                 "We don't have a transition frontier at the moment, so we're \
                  unable to verify any transactions."
         in
+
         let%bind diff' =
           O1trace.sync_thread "convert_transactions_to_verifiable" (fun () ->
               Or_error.try_with (fun () ->
                   Envelope.Incoming.map diff
                     ~f:
-                      (List.map ~f:(fun cmd ->
-                           User_command.to_verifiable ~ledger
-                             ~get:Base_ledger.get
-                             ~location_of_account:
-                               Base_ledger.location_of_account cmd
-                           |> Or_error.ok_exn ) ) ) )
+                      (Fn.compose
+                         snd (* remove the helper cache we folded with *)
+                         (List.fold_map ~init:F.Map.empty
+                            ~f:(fun running_cache cmd ->
+                              let verified_cmd =
+                                User_command.to_verifiable cmd
+                                  ~find_vk:(fun vk_hash account_id ->
+                                    (* a little "or else" operator *)
+                                    let ( |- ) x y_thunk =
+                                      match x with
+                                      | Error _ ->
+                                          y_thunk ()
+                                      | Ok z ->
+                                          Ok z
+                                    in
+                                    (* first we check if there's anything in the running cache within this diff so far *)
+                                    F.Map.find running_cache vk_hash
```

Comment:
Does it make sense to factor out (some of) this logic into a function in `Zkapp_command`? It seems like we probably want something similar to this in (at least) the staged ledger diff application logic too.